source/RoslynAnalyzers/AsyncAnalyzer.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Octopus.RoslynAnalyzers;
+
+using static Descriptors;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class AsyncAnalyzer : DiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+        VoidMethodsMustNotBeAsync,
+        MethodsReturningTaskMustBeAsync);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+        context.RegisterSyntaxNodeAction(CheckNode, SyntaxKind.MethodDeclaration);
+    }
+
+    void CheckNode(SyntaxNodeAnalysisContext context)
+    {
+        if (context.Node is not MethodDeclarationSyntax methodDec) return;
+
+        var isAsync = methodDec.Modifiers.Any(SyntaxKind.AsyncKeyword);
+        if (isAsync)
+        {
+            VoidMethods_MustNotBeAsync(context, methodDec);
+        }
+
+        if (methodDec.ReturnType is SimpleNameSyntax { Identifier.Text: "Task" or "ValueTask" })
+        {
+            MethodsReturningTask_MustBeAsync(context, methodDec, isAsync);
+        }
+    }
+
+    void VoidMethods_MustNotBeAsync(SyntaxNodeAnalysisContext context, MethodDeclarationSyntax methodDec)
+    {
+        if (methodDec.ReturnType is PredefinedTypeSyntax p && p.Keyword.IsKind(SyntaxKind.VoidKeyword))
+        {
+            context.ReportDiagnostic(Diagnostic.Create(VoidMethodsMustNotBeAsync, methodDec.Identifier.GetLocation()));
+        }
+    }
+
+    void MethodsReturningTask_MustBeAsync(SyntaxNodeAnalysisContext context, MethodDeclarationSyntax methodDec, bool isAsync)
+    {
+        // don't flag things that are async (that's good!) or abstract methods.
+        if (isAsync || methodDec.Modifiers.Any(SyntaxKind.AbstractKeyword)) return;
```

Comment:
Can we pull the `isAsync` out a level, and make line 38 an `else if`?
Or are you looking to keep this self contained (if so, there seems a bit of a mismatch in how it's implemented - the first one does the check in the caller, but the second one does the check itself)

Reply:
yeah that's a good idea.

What I was aiming for was a sort of hierarchy, with some basic high level checks, with detail in each method like this:

```
if(majorConcernOne) {
   check1()
   check2()
}

if(majorConcernTwo) {
   checkA()
   ...
}
```

In practice, it hasn't panned out as nicely as that. Thanks for the comment, I'll pull the isAsync out and simply not call this method if it's true, then it will align with the other method